File: medium/1306-jump-game-iii/solution.java
import java.util.ArrayDeque;
import java.util.Queue;

/**
 * 1306. Jump Game III
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public boolean canReach(int[] arr, int start) {
        int n = arr.length;
        boolean[] visited = new boolean[n];
        Queue<Integer> queue = new ArrayDeque<>();

        queue.offer(start);
        visited[start] = true;

        while (!queue.isEmpty()) {
            int index = queue.poll();

            if (arr[index] == 0) {
                return true;
            }

            int forward = index + arr[index];
            int backward = index - arr[index];

            if (forward < n && !visited[forward]) {
                visited[forward] = true;
                queue.offer(forward);
            }

            if (backward >= 0 && !visited[backward]) {
                visited[backward] = true;
                queue.offer(backward);
            }
        }

        return false;
    }
}

File: medium/1306-jump-game-iii/solution.py
from collections import deque
from typing import List


class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        """
        1306. Jump Game III
        Time: O(n)
        Space: O(n)
        """
        n = len(arr)
        visited = [False] * n
        queue = deque([start])
        visited[start] = True

        while queue:
            index = queue.popleft()

            if arr[index] == 0:
                return True

            for next_index in (index + arr[index], index - arr[index]):
                if 0 <= next_index < n and not visited[next_index]:
                    visited[next_index] = True
                    queue.append(next_index)

        return False
